### PR TITLE
feat: pivot table hide columns

### PR DESCRIPTION
**Pull request by IrakliJani**
Closes: #5031 

### Description:

Screenshot 2023-04-06 at 20 55 48

error boundary if you are hiding all metrics (0 rows = no data)

Screenshot 2023-04-06 at 20 57 17

### CODE REVIEW

File: packages/frontend/src/hooks/pivotTable/pivotQueryResults.ts [hunk 1]
```diff
@@ -56,10 +56,18 @@ export const pivotQueryResults = ({
...
-    const columnOrder = pivotConfig.columnOrder || [];
+
+    const hiddenFieldIds = pivotConfig.hiddenFieldIds || [];
+
+    const columnOrder = (pivotConfig.columnOrder || []).filter((id) => {
+        return !hiddenFieldIds.includes(id);
+    });
+
     // Headers (column index)
     const headerDimensions = pivotConfig.pivotDimensions.filter(
```
Comment (owlas) [inline packages/frontend/src/hooks/pivotTable/pivotQueryResults.ts:67]: hiding dimensions, will mean that data gets lost in the pivot, depending on the order of the data, the last value in the group of selected dimensions will be used and all the others will be discarded

File: packages/frontend/src/hooks/pivotTable/pivotQueryResults.ts [hunk 2]
```diff
@@ -99,9 +111,14 @@ export const pivotQueryResults = ({
         ...metricQuery.metrics,
         ...metricQuery.tableCalculations.map((tc) => tc.name),
     ]
+        .filter((m) => !hiddenFieldIds.includes(m))
```
Comment (owlas) [inline packages/frontend/src/hooks/pivotTable/pivotQueryResults.ts:114]: this makes sense

File: packages/frontend/src/hooks/pivotTable/pivotQueryResults.ts [hunk 3]
```diff
@@ -99,9 +111,14 @@ export const pivotQueryResults = ({
...
         .sort((a, b) => columnOrder.indexOf(a) - columnOrder.indexOf(b))
         .map((id) => ({ fieldId: id }));
 
+    if (metrics.length === 0) {
```
Comment (owlas) [inline packages/frontend/src/hooks/pivotTable/pivotQueryResults.ts:118]: good catch

File: packages/frontend/src/hooks/tableVisualization/useTableConfig.ts [hunk 4]
```diff
@@ -147,22 +155,33 @@ const useTableConfig = (
...
-            return pivotQueryResults({
-                pivotConfig: {
-                    pivotDimensions,
-                    metricsAsRows,
-                    columnOrder,
-                },
-                metricQuery: resultsData.metricQuery,
-                rows: resultsData.rows,
-            });
+
```
Comment (owlas) [inline packages/frontend/src/hooks/tableVisualization/useTableConfig.ts:158]: nice api 👍

**Review by owlas — CHANGES_REQUESTED** [review]
On reviewing this, we shouldn't allow the user to hide dimensions once pivoted, it means that data will be lost in the pivot.

e.g.

In this example if I pivot on `species` and hide `name`, then which animals do we keep? There will be a single metric value for `animal - bear` but will it be "Roger" or "gemima"? I can't think of a case where the user can be intentional about how this transformation would work, it would just be confusing.

It makes sense to hide metrics though because hiding a metric just hides all the metric values for that metric. But hiding a dimension randomly removes values from all metrics.

```
| kingdom | species | name   | count |
|---------|--------|--------|-------|
| animal  | bear   | roger  | 12    |
| animal  | bear   | gemima | 77    |
| animal  | bear   | suzie  | 9     |
| animal  | trout  | david  | 882   |
| animal  | trout  | jerry  | 11    |
| plant   | daisy  | penny  | 17    |
| plant   | daisy  | bruce  | 787   |
```

**Comment by IrakliJani** [comment]
@owlas @TuringLovesDeathMetal added a video demo to demonstrate what we discussed about hiding columns.

⚠️⚠️⚠️ note that this is now enabled for both old pivot table and the new one

here's a video: https://cln.sh/MgwBkfC0

**Comment by TuringLovesDeathMetal** [comment]
niiiiice! Looks really great! 

The only thing that I noticed which was a bit weird was when you pivoted the table and had a dimension hidden, the dimension was shown, but it was still greyed out (as though it was hidden). 

Can we make it so that it's not greyed out? (at 2:03) - otherwise, it feels weird that I'm forced to include the column, but I can't edit it

**Comment by IrakliJani** [comment]
@TuringLovesDeathMetal good catch! changing now

File: packages/frontend/src/hooks/pivotTable/pivotQueryResults.ts [hunk 5]
```diff
@@ -56,10 +56,18 @@ export const pivotQueryResults = ({
...
+    const columnOrder = (pivotConfig.columnOrder || []).filter((id) => {
+        return !hiddenFieldIds.includes(id);
+    });
+
     // Headers (column index)
     const headerDimensions = pivotConfig.pivotDimensions.filter(
-        (pivotDimension) => metricQuery.dimensions.includes(pivotDimension),
+        (pivotDimension) =>
+            metricQuery.dimensions.includes(pivotDimension) &&
+            !hiddenFieldIds.includes(pivotDimension),
```
Comment (owlas) [inline packages/frontend/src/hooks/pivotTable/pivotQueryResults.ts:70]: This still allows pivoted rows to be hidden

Comment (IrakliJani) [reply packages/frontend/src/hooks/pivotTable/pivotQueryResults.ts]: updated!

**Review by owlas — CHANGES_REQUESTED** [review]
If I hide a dimension and then pivot on another dimension, the dimension still remains hidden. I think there's an effect missing from `useTableConfig` or an error state from the table

Screenshot 2023-04-17 at 14 10 48

**Pull request merged**